# MiRNA biomarkers of pituitary adenoma recurrence

## Directory structure
```
├── Makefile
├── build — files, generated by the code from src
│   ├── downloads — downloaded files (e.g. library zips)
│   ├── lib — external executables used
│   ├── trimmed 
│   ├── revcomp
│   ├── short_long
│   ├── genome — reads, mapped to human genome reference
│   │   ├── alignment
│   │   ├── counts
│   │   └── index
│   ├── mature — reads, mapped to mature miRNA reference
│   │   ├── alignment
│   │   ├── counts
│   │   ├── index
│   ├── out — counts tables
│   │   ├── genome_counts.tsv
│   │   └── mirna_counts.tsv
│   └── multiqc_report.html — quality check report
├── data — input for the program, needs to be filled manually
└── src — source code
    ├── all_in_one.sh — all the scripts from src/pipeline combined, better not to use
    ├── pipeline — many small scripts with main logic
    │   ├── genome
    │   │   ├── align.sh
    │   │   ├── combine_counts.sh
    │   │   ├── get_counts.sh
    │   │   ├── index.sh
    │   │   ├── reheader.sh
    │   │   ├── sort.sh
    │   │   └── tag.sh
    │   ├── mature
    │   │   ├── align.sh
    │   │   ├── combine_counts.sh
    │   │   ├── get_counts.sh
    │   │   ├── index.sh
    │   │   └── sort.sh
    │   ├── revcomp.sh
    │   ├── setup
    │   │   ├── check_data.sh
    │   │   ├── check_tools.sh
    │   │   ├── download_genome_libs.sh
    │   │   └── setup_tools.sh
    │   ├── short_long_split.sh
    │   ├── trim.sh
    │   └── utils
    │       ├── combine_counts.R
    │       └── get_counts.py
    ├── requirements.txt — required libraries, gets called in setup_tools.sh
    └── run.sh — a script, which calls all scripts from src/pipeline

```

## Setup
### Data
The following `data` directory structure is required for the scripts to run:
```
data
├── adapters_rev_compl.fa
├── fastq
│   ├── N18_S2_L001_R2_001.fastq.gz
│   ├── N20_S3_L001_R2_001.fastq.gz
│   ├── N22_S4_L001_R2_001.fastq.gz
│   ├── N27_S5_L001_R2_001.fastq.gz
│   ├── N31_S6_L001_R2_001.fastq.gz
│   ├── N37_S7_L001_R2_001.fastq.gz
│   ├── N3_S1_L001_R2_001.fastq.gz
│   ├── N47_S8_L001_R2_001.fastq.gz
│   ├── N48_S9_L001_R2_001.fastq.gz
│   ├── N51_S10_L001_R2_001.fastq.gz
│   ├── N58_S11_L001_R2_001.fastq.gz
│   ├── N70_S12_L001_R2_001.fastq.gz
│   └── N74_S13_L001_R2_001.fastq.gz
├── hsa-genome-miRBase22v-onlymiRNAs-convforTagBAM.bed
├── mirnas.txt
└── ncbitochrom.txt
```
Contact me for a zip with it.

### Tools
It's better to create and activate a conda environment before running the scripts.
Also, the bioconda channel should be added to the conda installation:
```conda config --add channels bioconda```

No further preliminary setup is needed. The scripts may request additional manual actions at runtime.

## Run options
### all_in_one.sh
It can be executed, but it's better to be regarded only as a reference. It's just copy-pasted scripts from `src/pipeline`. 
Should be executed with `src` as a working directory.

### run.sh
Invokes all the scripts from `src/pipeline` at once. Should also be executed with `src` as a working directory.

### Makefile
This is the preferred run method. 
It invokes the scripts from `src/pipeline`, but unlike other options, 
it considers already built artifacts, so files are not overwritten and no computations are made twice.
It should be run from the project root as:
```
make all
```

Also, it provides a possibility to create only certain artifacts. For example:
```
make build/out/mirna_counts.tsv
```

Install some of the tools:
```
make setup_tools
```

Remove all build files except downloads:
```
make clean
```

Remove the build directory:
```
make clean_all
```
